Archiving system runs when new log file is created #390

awardle opened the pull request:
If archiving mode is set to date when a new log file is created, due to naming rules, the program will delete any
out of date archives.

Summary of changes
--------------------------
* Moved code to delete old archives in date mode to separate function 
* Moved code to get file name pattern into a separate function
* Added code to write function that runs the delete archive function when a new log file is started
* Added unit test for bug 390

Unit test for bug 390 is commented out as it takes 5 minutes to run it. This is due to the fact the minimum time period for archiving is a minute.

Fixes bug 390

awardle commented:
There is currently a bug in the patch which is causing 1 too many files to be deleted by the archiver sometimes.

ilya-g commented on tests/NLog.UnitTests/Targets/FileTargetTests.cs:
```diff
@@ -562,6 +562,71 @@ public void DeleteArchiveFilesByDate()
+        public void DeleteArchiveFilesByDateWithDateName()
+        {
+            var tempPath = Path.Combine(Path.GetTempPath(), Guid.NewGuid().ToString());
+            var tempFile = Path.Combine(tempPath, "${date:format=yyyyMMddHHmm}.txt");
```
You can use the format `yyyMMddHHmmss` to have new file created every second, so this test would require 5 seconds to run instead of 5 minutes.

ilya-g commented on src/NLog/Targets/FileTarget.cs:
```diff
@@ -630,6 +630,18 @@ protected override void Write(LogEventInfo logEvent)
+            if (this.ArchiveNumbering == ArchiveNumberingMode.Date && this.ArchiveEvery != FileArchivePeriod.None)
+            {
+                FileInfo fileInfo = new FileInfo(fileName);
+                if (!fileInfo.Exists)
```
Isn't it quite expensive to check whether the file exists on every write?

304NotModified replied:
Good catch @ilya-g

awardle commented:
Function now checks if the filename of the log file has changed since the last write to check if it is a new log file instead checking if the file exists. I think this should be less expensive than checking if the file exists.

304NotModified commented on src/NLog/Targets/FileTarget.cs:
```diff
@@ -1160,20 +1188,7 @@ private void DoAutoArchive(string fileName, LogEventInfo eventInfo)
-                //(2) User supplied the normal filename
-                fileNamePattern = this.ArchiveFileName.Render(eventInfo);
-            }
+            string fileNamePattern = GetFileNamePattern(fileName, eventInfo, fi);
```
Please no 'fi' as name

304NotModified commented on src/NLog/Targets/FileTarget.cs:
```diff
@@ -1210,6 +1225,25 @@ private void DoAutoArchive(string fileName, LogEventInfo eventInfo)
             }
         }
 
+        private string GetFileNamePattern(string fileName, LogEventInfo eventInfo, FileInfo fi)
```
Why do we have both filename and fileinfo as input?

awardle commented:
I've improved the documentation of the functions that I created and renamed some variables. In addition I have removed the fileInfo parameter from the GetFileNamePattern method.

304NotModified commented:
Looks good

ilya-g commented on src/NLog/Targets/FileTarget.cs:
```diff
@@ -630,6 +635,18 @@ protected override void Write(LogEventInfo logEvent)
+            // this log file and the archiving system is date/time based.
+            if (this.ArchiveNumbering == ArchiveNumberingMode.Date && this.ArchiveEvery != FileArchivePeriod.None)
+            {
+                if (this.previousFileName != fileName)
```
How would this work when the `FileName` contains not only the date layout but some other dynamic layout renderer, for example some client id or logger? Then this condition could be `true` almost on every write.

awardle replied:
Would it be better then if it stored in some sort of collection the previous log filenames instead of only the last log filename? Then the archiver cleaner (`DeleteOldDateArchive`) runs if the log file that is being written to is not in the collection.

awardle commented:
It now stores by default the last 20 file names used, although this can be changed in the configuration file, instead of just the last one. This will cause the `DeleteOldDateArchive` to run less often if the log file is changing between several different names.
